llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -7103,6 +7103,35 @@ Instruction *InstCombinerImpl::foldICmpCommutative(ICmpInst::Predicate Pred,
   if (Value *V = foldICmpWithLowBitMaskedVal(Pred, Op0, Op1, Q, *this))
     return replaceInstUsesWith(CxtI, V);
 
+  // Folding (X / Y) cmp X => X ~cmp 0 for some constant Y other than 0 or 1
+  {
+    Value *Dividend;
+    const APInt *Divisor;
+    if (match(Op0, m_UDiv(m_Value(Dividend), m_APInt(Divisor))) &&
+        Op1 == Dividend && !Divisor->isZero() && !Divisor->isOne()) {
+      return new ICmpInst(ICmpInst::getSwappedPredicate(Pred), Dividend,
+                          Constant::getNullValue(Dividend->getType()));
+    }
+
+    if (match(Op0, m_UDiv(m_Value(Dividend), m_APInt(Divisor))) &&
+        Op1 == Dividend && !Divisor->isZero() && !Divisor->isOne() &&
+        !ICmpInst::isUnsigned(Pred)) {
+      return new ICmpInst(ICmpInst::getSwappedPredicate(Pred), Dividend,
```

Comment:
This code seems unreachable :(

Reply:
Do you mean `m_SDiv`?

Reply:
Yes, looks like I copy-pasted the `match` arguments from the code above and forgot to change it to `sdiv`.